File: typescript/tests/owns_stdout.test.ts
/**
 * Stdout ownership: the ownsStdout declaration (effects contract §19.6).
 *
 * A command whose stdout IS the artifact declares it, and in machine mode the
 * envelope moves to stderr so the artifact's bytes are untouched. Outside
 * machine mode the declaration changes nothing at all.
 *
 * A handler's own document is a RAW stdout write, which app.test()'s captured
 * writers never see -- so the cases that pin the artifact's bytes drive
 * app.run() against patched process streams, the shape effects_exit_paths uses.
 */

import { strict as assert } from "node:assert";
import { test } from "node:test";
import {
	type App,
	createApp,
	defineMutatingCommand,
	defineReadOnlyCommand,
} from "../src/index.js";
import { dumpSchemaCore } from "../src/schema.js";
import { envelope } from "./envelope_helpers.js";

const DOC = '{"artifact":"v1"}';

/** Runs app.run(argv) against the real process streams, capturing both. */
async function captureRun(
	app: App,
	argv: string[],
): Promise<{ stdout: string; stderr: string }> {
	const out: string[] = [];
	const err: string[] = [];
	const realOut = process.stdout.write.bind(process.stdout);
	const realErr = process.stderr.write.bind(process.stderr);
	const patch =
		(sink: string[]) =>
		(chunk: string | Uint8Array): boolean => {
			sink.push(
				typeof chunk === "string" ? chunk : Buffer.from(chunk).toString(),
			);
			return true;
		};
	(process.stdout as unknown as { write: unknown }).write = patch(out);
	(process.stderr as unknown as { write: unknown }).write = patch(err);
	const savedExit = process.exitCode;
	try {
		await app.run(argv);
	} finally {
		(process.stdout as unknown as { write: unknown }).write = realOut;
		(process.stderr as unknown as { write: unknown }).write = realErr;
	}
	process.exitCode = savedExit;
	return { stdout: out.join(""), stderr: err.join("") };
}

function dumpApp(): App {
	const app = createApp({ name: "app", version: "1.0.0", help: "app" });
	app.command(
		defineReadOnlyCommand("dump", {
			help: "dump",
			ownsStdout: true,
			handler: () => {
				process.stdout.write(`${DOC}\n`);
				return 0;
			},
		}),
	);
	return app;
}

test("ownsStdout: outside machine mode the declaration changes nothing", async () => {
	const r = await captureRun(dumpApp(), ["dump"]);
	assert.equal(r.stdout, `${DOC}\n`);
	assert.equal(r.stderr, "");
});

test("ownsStdout: the document keeps stdout and the envelope moves to stderr", async () => {
	const r = await captureRun(dumpApp(), ["--json", "dump"]);
	assert.equal(r.stdout, `${DOC}\n`);
	const env = envelope(r.stderr);
	assert.equal(env.command, "dump");
	assert.equal(env.exit_code, 0);
});

test("ownsStdout: the diagnostics move with the envelope", async () => {
	const app = createApp({ name: "app", version: "1.0.0", help: "app" });
	app.command(
		defineReadOnlyCommand("dump", {
			help: "dump",
			ownsStdout: true,
			handler: (_args, ctx) => {
				ctx.info("wrote 1 row");
				ctx.warn("provisional");
				return 0;
			},
		}),
	);
	const r = await app.test(["--json", "dump"]);
	assert.equal(r.stdout, "");
	assert.deepEqual(envelope(r.stderr).diagnostics, [
		{ level: "info", message: "wrote 1 row" },
		{ level: "warn", message: "provisional" },
	]);
});

test("ownsStdout: an undeclared command keeps the envelope on stdout", async () => {
	const app = createApp({ name: "app", version: "1.0.0", help: "app" });
	app.command(
		defineReadOnlyCommand("plain", { help: "plain", handler: () => 0 }),
	);
	const r = await app.test(["--json", "plain"]);
	assert.equal(envelope(r.stdout).command, "plain");
	assert.equal(r.stderr, "");
});

test("ownsStdout: a preview's envelope moves too", async () => {
	const app = createApp({ name: "app", version: "1.0.0", help: "app" });
	app.command(
		defineMutatingCommand("dump", {
			help: "dump",
			ownsStdout: true,
			handler: (_args, ctx) => {
				ctx.effects.write("out.sql", "x");
				return 0;
			},
		}),
	);
	const r = await app.test(["--json", "--dry-run", "dump"]);
	assert.equal(r.stdout, "");
	const env = envelope(r.stderr);
	assert.equal(env.dry_run, true);
	assert.deepEqual(
		(env.preview as Record<string, unknown>[]).map((rec) => rec.verb),
		["write"],
	);
});

test("ownsStdout: a declared payload still rides the envelope on stderr", async () => {
	const app = createApp({ name: "app", version: "1.0.0", help: "app" });
	app.command(
		defineReadOnlyCommand("dump", {
			help: "dump",
			ownsStdout: true,
			payloadSchema: { type: "object" },
			handler: (_args, ctx) => {
				ctx.payload({ rows: 3 });
				return 0;
			},
		}),
	);
	const r = await app.test(["--json", "dump"]);
	assert.equal(r.stdout, "");
	assert.deepEqual(envelope(r.stderr).payload, { rows: 3 });
});

test("ownsStdout: the schema dump publishes the key only when true", () => {
	const app = createApp({ name: "app", version: "1.0.0", help: "app" });
	app.command(
		defineReadOnlyCommand("dump", {
			help: "dump",
			ownsStdout: true,
			handler: () => 0,
		}),
	);
	app.command(
		defineReadOnlyCommand("plain", { help: "plain", handler: () => 0 }),
	);
	const commands = dumpSchemaCore(
		app as unknown as Parameters<typeof dumpSchemaCore>[0],
	).commands as Record<string, Record<string, unknown>>;
	assert.equal((commands.dump as Record<string, unknown>).owns_stdout, true);
	assert.equal(
		"owns_stdout" in (commands.plain as Record<string, unknown>),
		false,
	);
});

File: go/strictcli/owns_stdout_test.go
package strictcli

import (
	"encoding/json"
	"fmt"
	"testing"
)

// Stdout ownership: the OwnsStdout declaration (effects contract §19.6).
//
// A command whose stdout IS the artifact declares it, and in machine mode the
// envelope moves to stderr so the artifact's bytes are untouched. Outside
// machine mode the declaration changes nothing at all.

const ownsStdoutDoc = `{"artifact":"v1"}`

func ownsStdoutApp(opts ...CmdOption) *App {
	app := NewApp("app", "1.0.0", "app")
	all := append([]CmdOption{WithEffect(EffectReadOnly), OwnsStdout()}, opts...)
	app.Command("dump", "dump", func(ctx *Context, kwargs map[string]interface{}) Outcome {
		fmt.Println(ownsStdoutDoc)
		return Exit(0)
	}, all...)
	return app
}

func TestOwnsStdoutChangesNothingOutsideMachineMode(t *testing.T) {
	r := ownsStdoutApp().Test([]string{"dump"})
	if r.ExitCode != 0 || r.Stdout != ownsStdoutDoc+"\n" || r.Stderr != "" {
		t.Fatalf("exit=%d stdout=%q stderr=%q", r.ExitCode, r.Stdout, r.Stderr)
	}
}

func TestOwnsStdoutMovesTheEnvelopeToStderr(t *testing.T) {
	r := ownsStdoutApp().Test([]string{"--json", "dump"})
	if r.Stdout != ownsStdoutDoc+"\n" {
		t.Fatalf("the artifact must own stdout byte-exactly, got %q", r.Stdout)
	}
	var env map[string]interface{}
	if err := json.Unmarshal([]byte(r.Stderr), &env); err != nil {
		t.Fatalf("stderr is not the envelope: %v (stderr=%q)", err, r.Stderr)
	}
	if env["command"] != "dump" || env["exit_code"].(float64) != 0 {
		t.Fatalf("envelope = %v", env)
	}
}

func TestOwnsStdoutMovesTheDiagnosticsWithTheEnvelope(t *testing.T) {
	app := NewApp("app", "1.0.0", "app")
	app.Command("dump", "dump", func(ctx *Context, kwargs map[string]interface{}) Outcome {
		ctx.Info("wrote 1 row")
		ctx.Warn("provisional")
		fmt.Println(ownsStdoutDoc)
		return Exit(0)
	}, WithEffect(EffectReadOnly), OwnsStdout())
	r := app.Test([]string{"--json", "dump"})
	if r.Stdout != ownsStdoutDoc+"\n" {
		t.Fatalf("stdout=%q", r.Stdout)
	}
	var env struct {
		Diagnostics []diagnosticRecord `json:"diagnostics"`
	}
	if err := json.Unmarshal([]byte(r.Stderr), &env); err != nil {
		t.Fatalf("stderr is not the envelope: %v", err)
	}
	want := []diagnosticRecord{
		{Level: "info", Message: "wrote 1 row"},
		{Level: "warn", Message: "provisional"},
	}
	if len(env.Diagnostics) != len(want) {
		t.Fatalf("diagnostics = %v", env.Diagnostics)
	}
	for i, d := range env.Diagnostics {
		if d != want[i] {
			t.Fatalf("diagnostics[%d] = %v, want %v", i, d, want[i])
		}
	}
}

func TestAnUndeclaredCommandKeepsTheEnvelopeOnStdout(t *testing.T) {
	app := NewApp("app", "1.0.0", "app")
	app.Command("plain", "plain", func(ctx *Context, kwargs map[string]interface{}) Outcome {
		return Exit(0)
	}, WithEffect(EffectReadOnly))
	r := app.Test([]string{"--json", "plain"})
	var env map[string]interface{}
	if err := json.Unmarshal([]byte(r.Stdout), &env); err != nil {
		t.Fatalf("stdout is not the envelope: %v (stdout=%q)", err, r.Stdout)
	}
	if r.Stderr != "" {
		t.Fatalf("stderr=%q", r.Stderr)
	}
}

func TestOwnsStdoutMovesAPreviewEnvelopeToo(t *testing.T) {
	app := NewApp("app", "1.0.0", "app")
	app.Command("dump", "dump", func(ctx *Context, kwargs map[string]interface{}) Outcome {
		ctx.Effects().Write("out.sql", "x")
		fmt.Println("-- sql")
		return Exit(0)
	}, WithEffect(EffectMutating), OwnsStdout())
	r := app.Test([]string{"--json", "--dry-run", "dump"})
	if r.Stdout != "-- sql\n" {
		t.Fatalf("stdout=%q", r.Stdout)
	}
	var env struct {
		DryRun  bool                     `json:"dry_run"`
		Preview []map[string]interface{} `json:"preview"`
	}
	if err := json.Unmarshal([]byte(r.Stderr), &env); err != nil {
		t.Fatalf("stderr is not the envelope: %v", err)
	}
	if !env.DryRun || len(env.Preview) != 1 || env.Preview[0]["verb"] != "write" {
		t.Fatalf("envelope = %+v", env)
	}
}

func TestOwnsStdoutIsPublishedBySchemaDumpOnlyWhenTrue(t *testing.T) {
	app := NewApp("app", "1.0.0", "app")
	app.Command("dump", "dump", func(ctx *Context, kwargs map[string]interface{}) Outcome {
		return Exit(0)
	}, WithEffect(EffectReadOnly), OwnsStdout())
	app.Command("plain", "plain", func(ctx *Context, kwargs map[string]interface{}) Outcome {
		return Exit(0)
	}, WithEffect(EffectReadOnly))
	schema := dumpSchemaCore(app)
	commands := schema["commands"].(map[string]interface{})
	dump := commands["dump"].(map[string]interface{})
	if dump["owns_stdout"] != true {
		t.Fatalf("dump entry = %v", dump)
	}
	plain := commands["plain"].(map[string]interface{})
	if _, present := plain["owns_stdout"]; present {
		t.Fatalf("the baseline must be omitted, plain entry = %v", plain)
	}
}
